File: 06.design-patterns/gift-selection/java/src/test/java/gift/ChildBuilder.java
package gift;

import java.util.ArrayList;
import java.util.List;

public class ChildBuilder {
    private final List<GiftRequest> giftRequests = new ArrayList<>();
    private Behavior behavior = Behavior.NICE;

    public static ChildBuilder aChild() {
        return new ChildBuilder();
    }

    public ChildBuilder nice() {
        this.behavior = Behavior.NICE;
        return this;
    }

    public ChildBuilder normal() {
        this.behavior = Behavior.NORMAL;
        return this;
    }

    public ChildBuilder naughty() {
        this.behavior = Behavior.NAUGHTY;
        return this;
    }

    public ChildBuilder requestingFeasibleGift(String giftName) {
        this.giftRequests.add(new GiftRequest(giftName, true));
        return this;
    }

    public ChildBuilder requestingInfeasibleGift(String giftName) {
        this.giftRequests.add(new GiftRequest(giftName, false));
        return this;
    }

    public Child build() {
        return new Child("Jane", "Doe", 9, this.behavior, this.giftRequests);
    }
}

File: 06.design-patterns/gift-selection/ts/tests/childBuilder.ts
import {Behavior, Child} from "../src/child";
import {GiftRequest} from "../src/giftRequest";

export class ChildBuilder {
    private behavior: Behavior = "nice";
    private readonly giftRequests: GiftRequest[] = [];

    static aChild(): ChildBuilder {
        return new ChildBuilder();
    }

    nice(): this {
        this.behavior = "nice";
        return this;
    }

    normal(): this {
        this.behavior = "normal";
        return this;
    }

    naughty(): this {
        this.behavior = "naughty";
        return this;
    }

    requestingFeasibleGift(giftName?: string): this {
        this.giftRequests.push(new GiftRequest(giftName ?? "A feasible gift", true));
        return this;
    }

    requestingInfeasibleGift(giftName?: string): this {
        this.giftRequests.push(new GiftRequest(giftName ?? "An infeasible gift", false));
        return this;
    }

    build(): Child {
        return new Child(
            "Jane",
            "Doe",
            9,
            this.behavior,
            this.giftRequests
        );
    }
}